File: solutions/python/triangle/1/triangle.py
def equilateral(sides):
    return is_valid_triangle(sides) and all(side == sides[0] for side in sides)

def isosceles(sides):
    side1, side2, side3 = sides
    return is_valid_triangle(sides) and ((side1 == side2 or side1 == side3) or (side2 == side3))

def scalene(sides):
    side1, side2, side3 = sides
    return is_valid_triangle(sides) and side1 != side2 and side1 != side3 and side2 != side3 

def is_valid_triangle(sides) -> bool:
    side1, side2, side3 = sides
    valid_sides = [
        (side1 + side2 >= side3),
        (side2 + side3 >= side1),
        (side1 + side3 >= side2)
    ]
    return len(sides) == 3 and all(sides) > 0 and all(valid_sides)

File: solutions/javascript/triangle/1/triangle.js
export class Triangle {
  constructor(side1, side2, side3) {
    this.side1  = side1;
    this.side2  = side2;
    this.side3  = side3;
    this.sides  = [this.side1, this.side2, this.side3];
  }
  get isEquilateral() {
    let greaterThanZero = sideLength => sideLength > 0;
    if (!this.sides.every(greaterThanZero)) {
      return false;
    } else if (this.side1 + this.side2 < this.side3 || this.side1 + this.side3 < this.side2 || this.side2 + this.side3 < this.side1) {
      return false;
    } else if (this.side1 === this.side2 && this.side2 === this.side3) {
      return true;
    } else {
      return false;
    }
  }
  get isIsosceles() {
    let greaterThanZero = sideLength => sideLength > 0;
    if (!this.sides.every(greaterThanZero)) {
      return false;
    } else if (this.side1 + this.side2 < this.side3 || this.side1 + this.side3 < this.side2 || this.side2 + this.side3 < this.side1) {
      return false;
    } else if (this.side1 === this.side2 || this.side1 === this.side3 || this.side2 === this.side3) {
      return true;
    } else {
      return false;
    }
  }
  get isScalene() {
    let greaterThanZero = sideLength => sideLength > 0;
    if (!this.sides.every(greaterThanZero)) {
      return false;
    } else if (this.side1 + this.side2 < this.side3 || this.side1 + this.side3 < this.side2 || this.side2 + this.side3 < this.side1) {
      return false;
    } else if (this.side1 !== this.side2 && this.side2 !== this.side3) {
      return true;
    } else {
      return false;
    }
  }
}
